Wizard-Framework for the TMS

klees opened the pull request:
I factored out some of the mechanism in `TMS/Booking` to get a generalized mechanism for different kind of wizards as a preperation for the creation mechanism for trainings we are about to create.

On the way to achieve this I also fixed a lot of tests. Unit tests are now running completely. The existing usages of the removed `Booking\Player` were refactored to the new Wizard.

After this is merged, these two tasks need to be done:

* Change construction of `CourseActions` according to new constructor of `CourseActionImpl`.
* Change `g_user_id` to `current_user_id` according to update in constructor of `CourseActionImpl`.

klees committed:
fixed tests for CourseActionImpl.

klees committed:
fixed typos in TableRelations

klees committed:
uninitialized var in ViewControl\Renderer

klees committed:
fixes in tests for CourseAction

klees committed:
fixed tests for CourseInfoHelper

klees committed:
fixed test

klees committed:
fixed uninitialized fields in Tables\Query.

klees committed:
UI: added empty rules in viewcontrol and table

klees committed:
UI: fixed test for YAML-parser

klees committed:
UI: fixed Panel test for new HTML

klees committed:
UI: added missing (and empty) context for glyph briefcase.

klees committed:
UI: fixed Deck-test

klees committed:
UI: fixed Card-Test

klees committed:
UI: fixed Button-tests and related tests

klees committed:
UI: fixed Panel-tests

klees committed:
TMS/Workflow: defined type for state

klees committed:
corrected Copyright

klees committed:
TMS/Workflow: added player for workflow

klees committed:
TMS/Workflow: streamlining in appendTo-methods of Step

klees committed:
TMS/Booking: implemented adapter from booking step to workflow step

klees committed:
TMS: Workflow -> Wizard

klees committed:
TMS/Booking: implemented Wizard

klees committed:
fixed label issue in sortation

klees committed:
TMS/Wizard: player does not need to be abstract.

klees committed:
TMS/Booking: component_class parameter for wizard

klees committed:
Booking/Wizard: no need to be abstract anymore

klees committed:
TMS/Booking: refactored Services according to new Wizard-mechanism

klees commented:
@dkloepfer Ich hab Änderungen in den TableRelations gemacht. Bitte einmal drübergucken. Danke!

klees commented:
@shecken Bitte guck durchgucken. Habe einiges in den ganzen Booking-Geschichten aufgeräumt und geändert.

klees committed:
TMS/Wizard: introduced SessionStateDB

klees committed:
TMS/Cancel: use new wizard mechanism

klees committed:
TMS/Booking: use component class for wizard id.

dkloepfer commented on src/TMS/TableRelations/Tables/Query.php:
```diff
@@ -8,15 +8,34 @@
+		$this->requested = null;
+		$this->group_by = array();
+		$this->filter = null;
+		$this->root_table;
```
Das dann auch = null?

dkloepfer reviewed (approved):
Ich sehe nichts radikales. Nehme an die tests laufen noch, also von mir aus approved.

klees committed:
TMS/TableRelations: missing null assignment

shecken commented on src/TMS/Booking/Wizard.php:
```diff
@@ -0,0 +1,155 @@
+require_once(__DIR__."/../../../Services/Form/classes/class.ilFormSectionHeaderGUI.php");
+
+use CaT\Ente\ILIAS\ilHandlerObjectHelper;
+use ILIAS\TMS\Wizard as W;
```
Wollten wir nicht Aussagekräftige Namen benutzen? Ich weiß das Wizard\Wizard doof aussieht, aber ich fänds trotzdem besser.

klees replied:
Ich kann `Wizard` hier nicht nutzen, weil das schon belegt ist (von der Klasse selber). Anderer Vorschlag?

klees replied:
`Wiz`?

shecken replied:
Ok

shecken commented on src/TMS/Wizard/Player.php:
```diff
@@ -0,0 +1,221 @@
+				->buildOverviewForm($state)
+				->getHtml();
+		}
+
```
Kriegen wir den Part nicht vorher raus? Finde das ist nicht die Aufgabe von runStep. Da fehlt irgendwie sowas wie prepareStep(), die dann diese Entscheidung trifft, was als nächstes passiert. Ich weiß das war vorher schon so, aber wenn wir da gerade Zeit investieren, dann kann man da ja mal gucken.
Da könnten dann auch die beiden ToDo's rein geabreitet werden.

shecken commented on Services/TMS/Booking/classes/ilTMSBookingGUI.php:
```diff
@@ -75,22 +77,46 @@ public function executeCommand() {
 		if((int)$this->g_user->getId() !== $usr_id && !$this->checkIsSuperiorEmployeeBelowCurrent($usr_id)) {
-			$this->redirectToPreviousLocation(array($this->g_lng->txt("no_permissions_to_book")), false);
+			$this->setParameter(null, null);
+			return $gui_bindings->redirectToPreviousLocation(array($this->g_lng->txt("no_permissions_to_book")), false);
```
Hm ich hätte jetzt in den bindings eher nur Dinge erwartet, die sich mit der Darstellung beschäftigen. Sprachvariable, unterschiedliche Formelemente oder so.

Redirect ist doch eher ne Art Tool um von A nach B zu kommen.

klees replied:
Fair enough. Für mich war `GUIBindings` die Bindings an ILIAS-GUI-Functionalitäten, zu denen Redirect imo schon gehört. Anderer Namensvorschlag?

shecken replied:
In ILIAS ist das eigentlich auch eine Funktion die von den GUI's benutzt wird. Das ganze wird erst zur Funktionalität indem man ilCtrl nur in GUI's verwenden darf.
Deshalb gibt es ja noch das Schlupfloch in den ilUtil mit denen man auch redirecten kann.

Ich würde das einfach aus den Bindings nehmen und in den Wizard verschieben. Die parents sollten ja auch da bekannt sein.

klees replied:
Das finde ich falsch. Der `Wizard` ist das "was", die `GUI-Bindings` das "wie".

shecken commented on src/TMS/Booking/GUIBindings.php:
```diff
@@ -0,0 +1,125 @@
+	/**
+	 * @inheritdocs
+	 */
+	protected function txt($id) {
```
Ich mag das hier irgendwie immer noch nicht. Klar sparen wir dadurch an public functionen, aber wenn wir jetzt schon eine eigene GUI dafür haben, dann sollte man mal Übergelgen ob "gui_binding->getPlayerTitle()" nicht für die Zukunft besser zu lesen ist, als "gui_binding->txt("title")" wo wir dann wieder gucken müssen zu was "title" umgemünzt wird.

Außerdem muss "aborted" auch noch umgearbeitet werden, da beim Abbrechen von Wartelistenbuchung oder FK-Buchung eine andere Meldung erscheinen muss. Dann bleiben wieder nur die Buttonlabel.

klees replied:
Ich verstehe deinen Punkt. Auf der anderen Seite: Wollen wir für jeden Wunsch analog zu "hier Titel anders" eine neue Methode einführen? Das scheint mir eher unlesbar. Gibt es vllt. einen Zwischenweg, z.B. Konstanten für die verwendeten Sprach-IDs einführen? Das ist ja im Grunde ein allgemeines Problem, was wir immer wieder bekommen werden, wenn wir solche Arten von Klassen ("generalisierte GUI-Klassen") haben.

shecken replied:
Das Problem ist ja weniger das Ding wo die Ausgabe nun her kommt. Es is dieser blöde Switch zwischen den GUI's. Mit den Bindings ist er ja nun nur an eine Stelle gewandert.

Dadurch das die Ausgabe nun in den Bindings sind, und hier die Werte ja eh gesetzt werden, finde ich getter dafür nicht so unübersichtlich, bzw. später schneller wieder zu finden.

Wie bekommen wir denn die Konstanten aus den Playern in die Bindings? Das bräuchten wir ja für jeden Fall eigene Bindings. Also WaitingBinding, SelfBookBinding, SuperiorBookBinding....

shecken commented on src/TMS/Booking/StepAdapter.php:
```diff
@@ -0,0 +1,86 @@
+
+namespace ILIAS\TMS\Booking;
+
+use ILIAS\TMS\Wizard as W;
```
Hier wieder dieser nichtssagende Alias.

shecken commented on src/TMS/Booking/StepAdapter.php:
```diff
@@ -0,0 +1,86 @@
+	 */
+	protected $user_id;
+
+	public function __construct(Step $wrapped, $crs_ref_id, $user_id) {
```
Was soll denn $wrapped für ein Step sein?
Und hier fehlt noch der Docstring für $crs_ref_id und $user_id.

klees replied:
`$wrapped` ist ein `Booking\Step`, sozusagen ein alter, der wird über diese Klasse zu einem `Wizard\Step`. Wie kann ich das klarer machen? Docstring füge ich hinzu.

shecken commented on src/TMS/CourseActionImpl.php:
```diff
@@ -31,21 +31,25 @@
 	 * @param int[] 	$contexts
 	 */
-	public function __construct(Ente\Entity $entity, \ilObject $owner, $priority, array $contexts) {
+	public function __construct(Ente\Entity $entity, \ilObject $owner, \ilObjUser $current_user, $priority, array $contexts) {
```
Wieso den ganzen user, wenn wir davon nur die ID brauchen? Ist current user == global user?

klees replied:
Ja, ich hab auf diese Art da eine versteckte Dependency auf `$DIC->user()` entfernt. Im Grunde ist dass hier einfach eine billige Übersetzung, ohne groß drüber nachzudenken. Die Frage ist also berechtigt. Pro wäre: Ich weiß so innen, das ich bestimmt die ID eines Users hab. Contra ist, dass wir das nirgendwo anders so machen.

shecken replied:
Na wer weiß was wir damit demnächst machen sollen/wollen. Von daher finde ich es nicht so schlimm. Vielleicht nur noch in den Docstring rein, das es der global user ist. Dann ist die Frage auch geklärt.

klees replied:
done

shecken reviewed (commented):
Ein paar Fragen und Anregungen.

klees committed:
TMS/Booking: added missing doc

klees committed:
TMS/Wizard: some more refactoring in Player

klees committed:
TMS/Booking: removed includes of outdated ilTMSBookingPlayerStateDB

klees committed:
TMS/Wizard: more intelligle namespace usage

klees committed:
TMS/CourseAction: some more docu

klees committed:
TMS/Booking: corrected class name

klees committed:
TMS/Wizard: GUIBindings -> ILIASBindings

klees committed:
TMS: respect new CourseActionImpl

klees commented:
Changed the instantiation of the one CourseAction I found in the ILIAS-base: `ToCourse`. The id of the global user was not used there.

Also changed the two Plugins that currenly use CourseActionImpl in the two PRs mentioned above.
